packages/schedule/src/Schedule.js

```diff
@@ -3,174 +3,327 @@
 const localRequestAnimationFrame =
   typeof requestAnimationFrame === 'function'
     ? requestAnimationFrame
-    : (undefined: any);
+    : undefined;
 const localCancelAnimationFrame =
-  typeof cancelAnimationFrame === 'function'
-    ? cancelAnimationFrame
-    : (undefined: any);
-
-const hasNativePerformanceNow =
-  typeof performance === 'object' && typeof performance.now === 'function';
+  typeof cancelAnimationFrame === 'function' ? cancelAnimationFrame : undefined;
+
+let getCurrentTime: () => number;
+
+// requestAnimationFrame does not run when the tab is in the background. If
+// we're backgrounded we prefer for that work to happen so that the page
+// continues to load in the background. So we also schedule a 'setTimeout' as
+// a fallback.
+// TODO: Need a better heuristic for backgrounded work.
+const ANIMATION_FRAME_TIMEOUT = 100;
+let rAFID;
+let rAFTimeoutID;
+const requestAnimationFrameWithTimeout = callback => {
+  // schedule rAF and also a setTimeout
+  rAFID = localRequestAnimationFrame(timestamp => {
+    // cancel the setTimeout
+    localClearTimeout(rAFTimeoutID);
+    callback(timestamp);
+  });
+  rAFTimeoutID = localSetTimeout(() => {
+    // cancel the requestAnimationFrame
+    localCancelAnimationFrame(rAFID);
+    callback(getCurrentTime());
+  }, ANIMATION_FRAME_TIMEOUT);
+};
 
-let now;
 if (hasNativePerformanceNow) {
   const Performance = performance;
-  now = function() {
+  getCurrentTime = function() {
     return Performance.now();
   };
 } else {
-  now = function() {
+  getCurrentTime = function() {
     return localDate.now();
   };
 }
 
-let scheduleWork: (
-  callback: FrameCallbackType,
-  options?: {timeout: number},
-) => CallbackIdType;
-let cancelScheduledWork: (callbackId: CallbackIdType) => void;
-
-if (!canUseDOM) {
-  const timeoutIds = new Map();
-
-  scheduleWork = function(
-    callback: FrameCallbackType,
-    options?: {timeout: number},
-  ): CallbackIdType {
-    // keeping return type consistent
-    const callbackConfig = {
-      scheduledCallback: callback,
-      timeoutTime: 0,
-      next: null,
-      prev: null,
-    };
-    const timeoutId = localSetTimeout(() => {
-      callback({
-        timeRemaining() {
-          return Infinity;
-        },
-        didTimeout: false,
-      });
-    });
-    timeoutIds.set(callback, timeoutId);
-    return callbackConfig;
+let requestCallback;
+let cancelCallback;
+let getTimeRemaining;
+
+if (typeof window === 'undefined') {
+  // If this accidentally gets imported in a non-browser environment, fallback
+  // to a naive implementation.
+  let timeoutID = -1;
+  requestCallback = (callback, absoluteTimeout) => {
+    timeoutID = setTimeout(callback, 0, true);
   };
-  cancelScheduledWork = function(callbackId: CallbackIdType) {
-    const callback = callbackId.scheduledCallback;
-    const timeoutId = timeoutIds.get(callback);
-    timeoutIds.delete(callbackId);
-    localClearTimeout(timeoutId);
+  cancelCallback = () => {
```

Comment:
Why arrow functions? I figured this would just become an ES file eventually?

Reply:
I mean ES5 file so that it can be published to npm directly.

Reply:
Ahh, when you use 'ES file', seems we all think it refer to 'ES6+ file'

Reply:
Should I get rid of const as well then?